File: web/src/entities/wallOfDeath.ts
import { ImmutableVec2 } from '../math.js';
import { lerp } from '../utils.js';
import { Entity } from '../entity.js';
import { PlayerSkill } from '../playerSkill.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import type { Animation } from '../media/animation.js';
import { ParticleSystem } from './particleSystem.js';

export class WallOfDeath extends Entity {
  private inited = false;
  private boost = false;
  private speed = 0;
  private frameCounter = 0;
  private running = false;
  private soundCountdown = 10;
  private readonly saw: Animation;
  private originalPosition = new ImmutableVec2();

  constructor() {
    super();
    this.setSize(new ImmutableVec2(10, 12 * 100));
    this.saw = Resource.getAnimation('data/images/saw.bmp');
  }

  getLayer(): number { return 1; }

  override onRespawn(): void {
    this.myRoom.destroyToTileRow(-1);
    this.setPosition(this.originalPosition);
    this.running = false;
    this.boost = false;
    this.soundCountdown = 10;
  }

  override onStartWallOfDeath(): void {
    if (!this.running) {
      this.running = true;
    } else {
      this.boost = true;
    }
  }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getDrawPositionX() + offsetX - this.getSize().x / 2);
    const y = Math.round(this.getDrawPositionY() + offsetY - this.getSize().y / 2);
    const saws = Math.floor(this.getSize().y / 12);

    for (let i = 0; i < saws; i++) {
      this.saw.drawFrame(ctx, Math.floor(this.frameCounter / 5) + i, x, y + i * 12);
    }
  }

  override update(): void {
    this.frameCounter++;

    if (!this.inited) {
      this.originalPosition = new ImmutableVec2(this.getPosition());
      this.inited = true;
    }

    if (!this.running) return;

    if (this.getPosition().x > 2950) {
      const ps = new ParticleSystem(this.saw, 2, 100, 20, 10, 20, 1, new ImmutableVec2(0, -50), 4.0);
      const x = Math.round(this.getDrawPositionX() - this.getSize().x / 2);
      const y = Math.round(this.getDrawPositionY() - this.getSize().y / 2);
      const saws = Math.floor(this.getSize().y / 12);

      for (let i = 0; i < saws; i++) {
        ps.setPositionWithSpread(new ImmutableVec2(x, y + i * 12), 0.1, true);
      }
      this.myRoom.addEntity(ps);
      this.remove();
      return;
    }

    for (let xo = 0; xo < 3; xo++) {
      const xt = xo + Math.floor((this.getPosition().x - this.getSize().x / 2) / this.myRoom.getTileWidth());
      this.myRoom.destroyToTileRow(xt - 1);
    }

    const heroX = this.myRoom.getHero().getPosition().x;
    const x = this.getPosition().x;
    const distance = heroX - x;

    if (distance > 190 && !this.boost) {
      this.setPosition(new ImmutableVec2(heroX - 190, this.getPosition().y));
    }

    if (distance < 80) this.boost = false;

    if (this.boost) {
      this.speed = lerp(5, 9, PlayerSkill.get());
    } else {
      const wantedSpeed = lerp(1, 1.9, PlayerSkill.get());
      this.speed = lerp(this.speed, wantedSpeed, 0.05);
    }

    this.soundCountdown -= this.speed;
    if (this.soundCountdown <= 0) {
      this.soundCountdown += 10;
      Sound.playSample('data/sounds/damage');
    }

    this.setPosition(new ImmutableVec2(this.getPosition().x + this.speed, this.getPosition().y));

    if (this.myRoom.getHero().Collides(this)) {
      this.myRoom.getHero().kill();
    }
  }
}

File: wii/src/entities/wall_of_death.cpp
#include "entities/wall_of_death.h"
#include "entities/particle_system.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "room.h"
#include "player_skill.h"
#include "utils.h"
#include <cmath>

WallOfDeath::WallOfDeath() {
    setSize(Vec2(10, 12 * 100));
    saw_ = Resource::getAnimation("data/images/saw.bmp");
}

void WallOfDeath::onRespawn() {
    room_->destroyToTileRow(-1);
    setPosition(originalPosition_);
    running_ = false;
    boost_ = false;
    soundCountdown_ = 10.0f;
}

void WallOfDeath::onStartWallOfDeath() {
    if (!running_) {
        running_ = true;
    } else {
        boost_ = true;
    }
}

void WallOfDeath::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(getDrawPositionX() + offsetX - size_.x / 2.0f));
    int y = static_cast<int>(std::round(getDrawPositionY() + offsetY - size_.y / 2.0f));
    int saws = static_cast<int>(size_.y) / 12;

    for (int i = 0; i < saws; i++) {
        saw_->drawFrame(renderer, wallFrameCounter_ / 5 + i, x, y + i * 12);
    }
}

void WallOfDeath::update() {
    wallFrameCounter_++;

    if (!inited_) {
        originalPosition_ = position_;
        inited_ = true;
    }

    if (!running_) return;

    if (position_.x > 2950.0f) {
        ParticleSystem* ps = new ParticleSystem(saw_, 2, 100, 20, 10.0f, 20.0f, 1, Vec2(0, -50), 4.0f);
        int x = static_cast<int>(std::round(getDrawPositionX() - size_.x / 2.0f));
        int y = static_cast<int>(std::round(getDrawPositionY() - size_.y / 2.0f));
        int saws = static_cast<int>(size_.y) / 12;

        for (int i = 0; i < saws; i++) {
            ps->setPositionWithSpread(Vec2(static_cast<float>(x), static_cast<float>(y + i * 12)), 0.1f, true);
        }
        room_->addEntity(ps);
        remove();
        return;
    }

    for (int xo = 0; xo < 3; xo++) {
        int xt = xo + static_cast<int>(std::floor((position_.x - size_.x / 2.0f) / room_->getTileWidth()));
        room_->destroyToTileRow(xt - 1);
    }

    float heroX = room_->getHero()->getPosition().x;
    float x = position_.x;
    float distance = heroX - x;

    if (distance > 190.0f && !boost_) {
        setPosition(Vec2(heroX - 190.0f, position_.y));
    }

    if (distance < 80.0f) boost_ = false;

    if (boost_) {
        speed_ = lerp(5.0f, 9.0f, PlayerSkill::get());
    } else {
        float wantedSpeed = lerp(1.0f, 1.9f, PlayerSkill::get());
        speed_ = lerp(speed_, wantedSpeed, 0.05f);
    }

    soundCountdown_ -= speed_;
    if (soundCountdown_ <= 0.0f) {
        soundCountdown_ += 10.0f;
        Sound::playSample("data/sounds/damage");
    }

    setPosition(Vec2(position_.x + speed_, position_.y));

    if (room_->getHero()->Collides(*this)) {
        room_->getHero()->kill();
    }
}
